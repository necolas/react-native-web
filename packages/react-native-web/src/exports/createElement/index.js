/**
 * Copyright (c) Nicolas Gallagher.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @noflow
 */

import AccessibilityUtil from '../../modules/AccessibilityUtil';
import createDOMProps from '../../modules/createDOMProps';
import React from 'react';
import { LocaleProvider } from '../../modules/useLocale';

const createElement = (component, props, options) => {
  // Use equivalent platform elements where possible.
  let accessibilityComponent;
  if (component && component.constructor === String) {
    accessibilityComponent =
      AccessibilityUtil.propsToAccessibilityComponent(props);
  }
  const Component = accessibilityComponent || component;
  const domProps = createDOMProps(Component, props, options);

  const element = React.createElement(Component, domProps);

  // Update locale context if element's writing direction prop changes
  const elementWithLocaleProvider = domProps.dir ? (
    <LocaleProvider
      children={element}
      direction={domProps.dir}
      locale={domProps.lang}
    />
  ) : (
    element
  );

  return elementWithLocaleProvider;
};

export default createElement;
