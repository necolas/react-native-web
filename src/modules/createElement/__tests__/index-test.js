/* eslint-env jasmine, jest */

import createElement from '..';
import { shallow, render } from 'enzyme';

describe('modules/createElement', () => {
  test('it renders different DOM elements', () => {
    let component = render(createElement('span'));
    expect(component).toMatchSnapshot();
    component = render(createElement('main'));
    expect(component).toMatchSnapshot();
  });

  test('it normalizes event.nativeEvent', done => {
    const onClick = e => {
      e.nativeEvent.timestamp = 1496876171255;
      expect(e.nativeEvent).toMatchSnapshot();
      done();
    };
    const component = shallow(createElement('span', { onClick }));
    component.find('span').simulate('click', {
      nativeEvent: {}
    });
  });

  describe('when ARIA role is "button"', () => {
    [{ disabled: true }, { disabled: false }].forEach(({ disabled }) => {
      describe(`and disabled is "${disabled}"`, () => {
        [{ name: 'Enter', which: 13 }, { name: 'Space', which: 32 }].forEach(({ name, which }) => {
          test(`"onClick" is ${disabled ? 'not ' : ''}called when "${name}" is pressed`, () => {
            const onClick = jest.fn();
            const component = shallow(
              createElement('span', { accessibilityRole: 'button', disabled, onClick })
            );
            component.find('span').simulate('keyPress', {
              isDefaultPrevented() {},
              nativeEvent: {},
              preventDefault() {},
              which
            });
            expect(onClick).toHaveBeenCalledTimes(disabled ? 0 : 1);
          });
        });
      });
    });
  });
});
