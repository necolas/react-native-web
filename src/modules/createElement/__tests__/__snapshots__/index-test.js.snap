// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/createElement it normalizes event.nativeEvent 1`] = `
Object {
  "_normalized": true,
  "changedTouches": Array [],
  "identifier": undefined,
  "locationX": undefined,
  "locationY": undefined,
  "pageX": undefined,
  "pageY": undefined,
  "preventDefault": [Function],
  "stopImmediatePropagation": [Function],
  "stopPropagation": [Function],
  "target": undefined,
  "timestamp": 1496876171255,
  "touches": Array [],
}
`;

exports[`modules/createElement it renders different DOM elements 1`] = `<span />`;

exports[`modules/createElement it renders different DOM elements 2`] = `<main />`;
